rules/no-unnecessary-polyfills.js

```diff
@@ -0,0 +1,173 @@
+
+const polyfills = Object.keys(compatData).map(feature => {
+	let [ecmaVersion, constructorName, methodName = ''] = feature.split('.');
+
+	if (ecmaVersion === 'es') {
+		ecmaVersion = `(${ecmaVersion}\\d*)`;
+	}
+
+	constructorName = `(${constructorName}|${camelCase(constructorName)})`;
+	if (methodName) {
+		methodName = `(${methodName}|${camelCase(methodName)})`;
+	}
+
+	const methodOrConstructor = methodName || constructorName;
+
+	return {
+		feature,
+		pattern: new RegExp(
+			`^((${prefixes}?`
+				+ `(${ecmaVersion}${delimiter}${constructorName}${delimiter}${methodName}|` // Ex: es6-array-copy-within
+				+ `${constructorName}${delimiter}${methodName}|` // Ex: array-copy-within
+				+ `${ecmaVersion}${delimiter}${constructorName})` // Ex: es6-array
+				+ `${suffixes}?)|`
+				+ `(${prefixes}${methodOrConstructor}|${methodOrConstructor}${suffixes})` // Ex: polyfill-copy-within / polyfill-promise
+				+ `${additionalPolyfillPatterns[feature] || ''})$`,
+			'i',
+		),
+	};
+});
+
+function checkVersion(nodeVersion, featureVersion) {
+	const supportedNodeVersion = semver.coerce(featureVersion);
+	const validTargetVersion = semver.minVersion(nodeVersion);
+	return validTargetVersion
+		? semver.lte(supportedNodeVersion, validTargetVersion)
+		: semver.ltr(supportedNodeVersion, nodeVersion);
+}
+
+function report(context, node, feature) {
+	let [ecmaVersion, namespace, method = ''] = feature.split('.');
+	if (namespace === 'typed-array' && method.endsWith('-array')) {
+		namespace = method;
+		method = '';
+	}
+
+	const delimiter = method && (ecmaVersion === 'node' ? '.' : '#');
+
+	context.report({
+		node,
+		messageId: MESSAGE_ID_POLYFILL,
+		data: {
+			featureName: `${constructorCase(namespace)}${delimiter}${camelCase(method)}`,
+		},
+	});
+}
+
+function processRule(context, node, moduleName, nodeVersion) {
+	if (!moduleName || typeof moduleName !== 'string') {
+		return;
+	}
+
+	const importedModule = moduleName.slice(0, -path.extname(moduleName).length);
+	const coreJsModuleFeatures = coreJsEntries[importedModule.replace('core-js-pure', 'core-js')];
+
+	if (coreJsModuleFeatures) {
+		if (coreJsModuleFeatures.length === 1) {
+			if (checkVersion(nodeVersion, compatData[coreJsModuleFeatures[0]].node)) {
+				report(context, node, coreJsModuleFeatures[0]);
+			}
+		} else if (coreJsModuleFeatures.every(feature => compatData[feature].node && checkVersion(nodeVersion, compatData[feature].node))) {
+			context.report({
+				node,
+				messageId: MESSAGE_ID_CORE_JS,
+				data: {
+					coreJsModule: moduleName,
+				},
+			});
+		}
+
+		return;
+	}
+
+	const polyfill = polyfills.find(({pattern}) => pattern.test(importedModule));
+	if (polyfill) {
+		const [, namespace, method = ''] = polyfill.feature.split('.');
+		const features = coreJsEntries.find(it => it === `core-js/actual/${namespace}${method && '/'}${method}`);
+		if (features.every(feature => compatData[feature].node && checkVersion(nodeVersion, compatData[feature].node))) {
```

Comment:
Does it make a difference? I think it's good right now. By the way I don't know where `setIntersection` comes from.

Reply:
I think that's it's better than only Node support. `setIntersection` is just an abstract name of function intersection of sets.

Reply:
I agree that it is better. Maybe `targets` could be used directly as an option instead of the current Node.js specific option (targetVersion)?

```md
### targets

Type: `string`

The `targets` option allows to specify the target versions. This option is a [Browserlist](https://github.com/browserslist/browserslist) query.
If the option is unspecified, the targets are taken from the `package.json`, in the `browserslist` field, or alternatively, in the `engines.node` field (`targets` will be set to `node {engines.node}`).

``js
"unicorn/no-unnecessary-polyfills": ["error", {	"targets": "node >=12" }]
``

``js
"unicorn/no-unnecessary-polyfills": ["error", {	"targets": ["node 14.1.0", "chrome 95"] }]
``

```

What do you think about this?

Reply:
As an option -) Note that `core-js-compat` supports not only Browserlist format, it's almost similar to Babel targets format with some extensions. You can just add a link to the format of `targets` option.

Reply:
Thanks you for your feedback :)

> The `targets` option allows to specify the target versions. This option could be [a Browserlist query](https://github.com/browserslist/browserslist) or a targets object, see https://github.com/zloirock/core-js/tree/master/packages/core-js-compat#targets-option for more informations.

?

Reply:
👍

Reply:
Unfortunately, browserlist does not support this syntax `>=4 <5.2 || >6.0.0` (semver):
``Unknown browser query `node >=4 <5.2 || >6.0.0`. Maybe you are using old Browserslist or made typo in query.``

Reply:
The upper bound makes no sense for polyfills. You can use, for example, `minVersion` from `semver` package if you wanna use `package.json` `node` field.

Reply:
> It's also unfortunately that only non semver-patch Node.js versions are "supported"

What do you mean?

Reply:
`require('core-js-compat/compat')({ targets: 'node 17.3.1' })` -> `Unknown version 17.3.1 of Node.js`

Reply:
Nevermind, v8 updates (and so new features) are always semver major or minor.

Reply:
You can use `require('core-js-compat/compat')({ targets: { node: '17.3.1' } })` instead of `browserslis` query. `browserslis` is too limited.

Reply:
Also, `require('core-js-compat').data['es.array.iterator'].node` is `'10.0'` (expected) while `require('core-js-compat/compat')({ targets: {node:'12'} }).list.includes('es.array.iterator')` is `false` (unexpected).
Shouldn't the function return the features of this version as well as those of previous versions?

`require('core-js-compat/compat')({ targets: 'node <= 12' }).list.includes('es.array.iterator')` is `true` which is also IMO strange because `es.array.iterator` is not available in `node<=10` and so should not be present in the result of this query.

Have I misunderstood the purpose of this function?

Reply:
`core-js-compat/compat` returns a list of modules that **required** for the target version, `es.array.iterator` is not required for NodeJS 12.

`node <= 12` on the `browserslist` side normalized to NodeJS 0.2, `es.array.iterator` is required for NodeJS 0.2.